model/test_results.go

```diff
@@ -0,0 +1,309 @@
+		t.env,
+		conf.Bucket.TestResultsBucket,
+		t.Artifact.Prefix,
+		string(pail.S3PermissionsPrivate),
+		true,
+	)
+	if err != nil {
+		return errors.Wrap(err, "problem creating bucket")
+	}
+
+	testNames := make([]string, len(results))
+	for i, result := range results {
+		testNames[i] = result.TestName
+		data, err := bson.Marshal(result)
+		if err != nil {
+			return errors.Wrap(err, "problem marshalling bson")
+		}
+
+		if err := bucket.Put(ctx, result.TestName, bytes.NewReader(data)); err != nil {
+			return errors.Wrap(err, "problem uploading test results to bucket")
+		}
+
+	}
+
+	return errors.Wrap(t.appendTestNames(ctx, testNames), "problem updating log metadata during upload")
+}
+
+func (t *TestResults) appendTestNames(ctx context.Context, testNames []string) error {
+	if t.env == nil {
+		return errors.New("cannot append with a nil environment")
+	}
+
+	if t.ID == "" {
+		t.ID = t.Info.ID()
+	}
+
+	set := bson.M{}
+	for _, testName := range testNames {
+		set[bsonutil.GetDottedKeyName(testResultsArtifactKey, testResultsArtifactInfoTestsKey, testName)] = false
+	}
+	updateResult, err := t.env.GetDB().Collection(testResultsCollection).UpdateOne(
+		ctx,
+		bson.M{"_id": t.ID},
+		bson.M{"$set": set},
+	)
+	grip.DebugWhen(err == nil, message.Fields{
+		"collection":   testResultsCollection,
+		"id":           t.ID,
+		"updateResult": updateResult,
+		"test_names":   testNames,
+		"op":           "append test names to a test results record",
+	})
+	if err == nil && updateResult.MatchedCount == 0 {
+		err = errors.Errorf("could not find test results record with id %s in the database", t.ID)
+	}
+
+	return errors.Wrapf(err, "problem appending test names to %s", t.ID)
+}
+
+func (t *TestResults) SetTestRecorded(ctx context.Context, testName string, recorded bool) error {
+	if t.env == nil {
+		return errors.New("cannot set test recorded with a nil environment")
+	}
+
+	if t.ID == "" {
+		t.ID = t.Info.ID()
+	}
+
+	updateResult, err := t.env.GetDB().Collection(testResultsCollection).UpdateOne(
+		ctx,
+		bson.M{"_id": t.ID},
+		bson.M{
+			"$set": bson.M{
+				bsonutil.GetDottedKeyName(testResultsArtifactKey, testResultsArtifactInfoTestsKey, testName): recorded,
+			},
+		},
+	)
+	grip.DebugWhen(err == nil, message.Fields{
+		"collection":   testResultsCollection,
+		"id":           t.ID,
+		"updateResult": updateResult,
+		"test_name":    testName,
+		"op":           fmt.Sprintf("set test name recorded to %v", recorded),
+	})
+	if err == nil && updateResult.MatchedCount == 0 {
+		err = errors.Errorf("could not find test results record with id %s in the database", t.ID)
+	}
+
+	return errors.Wrapf(err, "problem setting test recorded for %s", t.ID)
+}
+
+// TestResultsInfo describes information unique to a single task execution.
+type TestResultsInfo struct {
+	Project   string `bson:"project,omitempty"`
+	Version   string `bson:"version,omitempty"`
+	Variant   string `bson:"variant,omitempty"`
+	Distro    string `bson:"distro,omitempty"`
```

Comment:
seems odd to have this here

Reply:
it is used for the historical stats data